packages/perseus/src/widgets/interactive-graph.test.tsx

```diff
@@ -0,0 +1,182 @@
+        const guess: PerseusGraphType = {
+            type: "segment",
+            coords: [
+                [
+                    [99, 0],
+                    [1, 1],
+                ],
+            ],
+        };
+        const rubric: Rubric = createRubric({
+            type: "segment",
+            coords: [
+                [
+                    [0, 0],
+                    [1, 1],
+                ],
+            ],
+        });
+
+        const result = InteractiveGraph.widget.validate(guess, rubric, null);
+
+        expect(result).toEqual({
+            type: "points",
+            earned: 0,
+            total: 1,
+            message: null,
+        });
+    });
+
+    it("awards points if guess.coords is right", () => {
+        const guess: PerseusGraphType = {
+            type: "segment",
+            coords: [
+                [
+                    [0, 0],
+                    [1, 1],
+                ],
+            ],
+        };
+        const rubric: Rubric = createRubric({
+            type: "segment",
+            coords: [
+                [
+                    [0, 0],
+                    [1, 1],
+                ],
+            ],
+        });
+
+        const result = InteractiveGraph.widget.validate(guess, rubric, null);
+
+        expect(result).toEqual({
+            type: "points",
+            earned: 1,
+            total: 1,
+            message: null,
+        });
+    });
+
+    it("allows points of a segment to be specified in reverse order", () => {
+        const guess: PerseusGraphType = {
+            type: "segment",
+            coords: [
+                [
+                    [1, 1],
+                    [0, 0],
+                ],
+            ],
+        };
+        const rubric: Rubric = createRubric({
+            type: "segment",
+            coords: [
+                [
+                    [0, 0],
+                    [1, 1],
+                ],
+            ],
+        });
+
+        const result = InteractiveGraph.widget.validate(guess, rubric, null);
+
+        expect(result).toEqual({
+            type: "points",
+            earned: 1,
+            total: 1,
+            message: null,
+        });
+    });
+
+    it("does not modify the `guess` data", () => {
+        const guess: PerseusGraphType = {
+            type: "segment",
+            coords: [
+                [
+                    [1, 1],
+                    [0, 0],
+                ],
+            ],
+        };
+        const rubric: Rubric = createRubric({
+            type: "segment",
+            coords: [
+                [
+                    [0, 0],
+                    [1, 1],
+                ],
+            ],
+        });
+
+        InteractiveGraph.widget.validate(guess, rubric, null);
+
+        expect(guess.coords).toEqual([
+            [
+                [1, 1],
+                [0, 0],
+            ],
+        ]);
+    });
+
+    it("does not modify the `rubric` data", () => {
```

Comment:
I don't really understand this test.
1. Are the coords supposed to be inverted like in the test above?
2. Why would type not be "segment"?

Reply:
I should probably add a comment to clarify why this test is needed. But it fails if I replace the `deepClone` of `rubric.correct.coords` with a shallow-copy `slice`:

```ts
let guess = Util.deepClone(userInput.coords);
let correct = rubric.correct?.coords.slice();
```

The elements of `correct` later get sorted in-place. This test ensures that that sorting does not disturb the original rubric data.

Re: the `invariant(rubric.correct.type === "segment");` - that's needed for TS to understand that `rubric.correct` is always going to have `coords`. The type of `rubric.correct` is `PerseusGraphType`, so we have to narrow the type to just segment graphs.